File: 백준/Silver/2178. 미로 탐색/미로 탐색.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
    static int N, M;
    static int[][] arr;
    static boolean[][] visited;
    static int[] dx = {1, 0, -1, 0};
    static int[] dy = {0, -1, 0, 1};
    
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        arr = new int[N][M];
        visited = new boolean[N][M];

        for (int i = 0; i < N; i++) {
            String str = br.readLine();
            for (int j = 0; j < M; j++) {
                arr[i][j] = str.charAt(j) - '0';
            }
        }


        visited[0][0] = true;
        bfs(0, 0);

        System.out.println(arr[N - 1][M - 1]);

    }

    public static void bfs(int x, int y) {
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{x, y});

        while (!q.isEmpty()) {
            int now[] = q.poll();
            int nx = now[0];
            int ny = now[1];

            for (int i = 0; i < 4; i++) {
                int ox = nx + dx[i];
                int oy = ny + dy[i];

                if (ox >= 0 && oy >= 0 && ox < N && oy < M) {
                    if (!visited[ox][oy] && arr[ox][oy] != 0) {
                        q.add(new int[]{ox, oy});
                        arr[ox][oy] = arr[nx][ny] + 1;
                        visited[ox][oy] = true;
                    }
                }

            }
        }
    }
}

File: 백준/Silver/2178. 미로 탐색/미로 탐색.kt
import java.util.*

fun main() = with(System.`in`.bufferedReader()) {
    var st = StringTokenizer(readLine())
    val w = st.nextToken().toInt()
    val h = st.nextToken().toInt()

    val miro = Array(w) {
        IntArray(h)
    }

    for (i in 0..w - 1) {
        val str = readLine()
        for (j in 0..h - 1) {
            miro[i][j] = str[j] - '0'
        }
    }

    val q = LinkedList<IntArray>()
    q.add(intArrayOf(0, 0))
    while(!q.isEmpty()) {
        val now = q.poll()

        repeat(DIR_NUM) {
            val nx = now[0] + dx[it]
            val ny = now[1] + dy[it]
            if(nx >= 0 && ny >= 0 && nx < w && ny < h && miro[nx][ny] == 1) {
                miro[nx][ny] += miro[now[0]][now[1]]
                q.add(intArrayOf(nx, ny))
            }
        }
    }

    print(miro[w-1][h-1])

}

const val DIR_NUM = 4
val dx: Array<Int> = arrayOf(0, 1, 0, -1)
val dy: Array<Int> = arrayOf(1, 0, -1, 0)

fun bfs(land: Array<IntArray>, x: Int, y: Int) {
    land[x][y] = 0;

    repeat(DIR_NUM) {
        val nx = x + dx[it]
        val ny = y + dy[it]

        if (nx >= 0 && ny >= 0 && nx < land.size && ny < land[0].size && land[nx][ny] == 1) {
            land[nx][ny] = 0
            dfs(land, nx, ny)
        }
    }
}


fun dfs(land: Array<IntArray>, x: Int, y: Int) {
    land[x][y] = 0;

    repeat(DIR_NUM) {
        val nx = x + dx[it]
        val ny = y + dy[it]

        if (nx >= 0 && ny >= 0 && nx < land.size && ny < land[0].size && land[nx][ny] == 1) {
            land[nx][ny] = 0
            dfs(land, nx, ny)
        }
    }
}